server/aws-lsp-codewhisperer/src/language-server/amazonQServiceManager/AmazonQTokenServiceManager.ts

```diff
@@ -0,0 +1,403 @@
+            // TODO: do more validation if necessary, for now, checking only existence of the profile
+            // TODO: do we need to reset service here if requested profile does not exist anymore?
+            this.logging?.log(`Amazon Q Profile ${newProfileArn} is not valid`)
+            this.resetCodewhispererService()
+            this.state = 'PENDING_Q_PROFILE'
+
+            throw new AmazonQServiceInvalidProfileError('Requested Amazon Q Profile does not exist')
+        }
+
+        if (!this.activeIdcProfile) {
+            this.initializeCodewhispererService('identityCenter', newProfile.region)
+            this.activeIdcProfile = newProfile
+
+            return
+        }
+
+        // Profile didn't change
+        if (this.activeIdcProfile && this.activeIdcProfile.arn === newProfile.arn) {
+            // Update cached profile fields, keep existing client
+            this.logging?.log('Amazon Q Profile Change: profile selection did not change')
+            this.activeIdcProfile = newProfile
+            this.state = 'INITIALIZED'
+
+            return
+        }
+
+        // At this point new valid profile is selected.
+
+        const oldRegion = this.activeIdcProfile.region
+        const newRegion = newProfile.region
+        if (oldRegion === newRegion) {
+            this.logging?.log(`Amazon Q server: new profile is in the same region as old one, keeping exising service.`)
+            this.activeIdcProfile = newProfile
+            this.state = 'INITIALIZED'
+
+            return
+        }
+
+        this.logging?.log(`Amazon Q: switching service client region from ${oldRegion} to ${newRegion}`)
+
+        // Selected new profile is in different region. Re-initialize service and terminate inflight requests
+        this.resetCodewhispererService()
+        this.initializeCodewhispererService('identityCenter', newProfile.region)
+
+        this.activeIdcProfile = newProfile
+
+        return
+    }
+
+    public getCodewhispererService(): CodeWhispererServiceToken {
+        // Prevent initiating requests while profile change is in progress.
+        if (this.state === 'PENDING_Q_PROFILE_UPDATE') {
+            throw new AmazonQServicePendingProfileUpdateError()
+        }
+
+        this.handleSsoConnectionChange()
+
+        if (this.state === 'INITIALIZED' && this.cachedCodewhispererService) {
+            return this.cachedCodewhispererService
+        }
+
+        if (this.state === 'PENDING_CONNECTION') {
+            throw new AmazonQServicePendingSigninError()
+        }
+
+        if (this.state === 'PENDING_Q_PROFILE') {
+            throw new AmazonQServicePendingProfileError()
+        }
+
+        throw new AmazonQServiceNotInitializedError()
+    }
+
+    private resetCodewhispererService() {
+        // TODO: terminate inflight requests for old service, if exist
+
+        this.cachedCodewhispererService = undefined
+        this.activeIdcProfile = undefined
+        this.state = 'PENDING_CONNECTION'
```

Comment:
I do not think we should reset the state inside of this method. I think we should change to 'PENDING_CONNECTION' only on sign out.